size = as.integer(c(10^6,2*10^6,3*10^6,4*10^6,5*10^6)) # The library sizes of the supsampled reads
load("141103-1_GeneDescription_Osa") # Load gene description
at = read.table("Sample_attribute.txt", header = F, stringsAsFactors = F, sep = "\t") # Load sample information
at2 = NULL
rawcnt = NULL
for (n in size) {
  for(i in at$V2){
    tmp = read.table(sprintf("RSEM_genes_results/Index%03d_%s.genes.results",i,n), header = T, stringsAsFactors = F, sep = "\t")
    rawcnt = cbind(rawcnt,tmp$expected_count)
    at2 = rbind(at2,c(i,n))
  }
}
rownames(rawcnt) = tmp[,1] 

log2rpm = t(log2(t(rawcnt)/colSums(rawcnt[des$NormalizationGroup=="data",])*10^6+1))
mean.log2rpm = apply(log2rpm[des$NormalizationGroup=="data",], 1, mean)
hostgene = grep("Os[0-1][0-9]t", rownames(log2rpm))
log2rpm.expressed = log2rpm[hostgene,]
libs = unique(at$V3)
rawcnt.expressed = rawcnt[hostgene,]

#### Fig. 4 A,B
source("../Os2015/Os_edQTL/scripts/colname2rgb.R")
pdf("Fig4AB.pdf")
plot(log2rpm.expressed[,50], log2rpm.expressed[,49], pch = 16, col = colnames2rgb("black",15), main=cor(log2rpm.expressed[,50], log2rpm.expressed[,49])) # Fig. 4A
plot(log2rpm.expressed[,55], log2rpm.expressed[,49], pch = 16, col = colnames2rgb("black",15), main=cor(log2rpm.expressed[,55], log2rpm.expressed[,49])) # Fig. 4B
dev.off()

#### Fig. 4 C

## Because of the limitation for file size of the upload server, we cannnot upload xxxx.depth files generated by samtools. You can run this script by skkiping this commented out lines. 
# Random.cov = read.table("bam/random.depth", header = F, stringsAsFactors = F)
# Lasy.cov = read.table("bam/lasy-seq.depth", header = F, stringsAsFactors = F)
# colnames(Random.cov) = c("gene", "pos","depth")
# colnames(Lasy.cov) = c("gene", "pos","depth")
# gc()
# gc()
# save(Random.cov, Lasy.cov, file="depth")

Load("depth")
tmp = table(Lasy.cov[,1])
tmp2 = cumsum(tmp)
index = matrix(1,ncol = 2, nrow = length(tmp2))
rownames(index) = names(tmp2)
index[,2] = tmp2
index[2:nrow(index),1] = tmp2[1:length(tmp2)-1]+1

random.depth = rep(0,max(tmp[hostgene]))
lasy.depth = rep(0,max(tmp[hostgene]))

Random.cov = Random.cov[,3]
Lasy.cov = Lasy.cov[,3]
gc()
gc()
for(gn in rownames(log2rpm.expressed)){
  cat(gn)
  cat("\n")
  pos = index[gn,]
  random = rev(Random.cov[pos[1]:pos[2]])
  lasy = rev(Lasy.cov[pos[1]:pos[2]])
  random.depth[1:length(random)] = random.depth[1:length(random)] + random
  lasy.depth[1:length(lasy)] = lasy.depth[1:length(lasy)] + lasy
}
Random.cov = data.frame(pos = 1:4000, depth = random.depth[126:4125]) # Because add 125 nt of As in bowtie-build
Lasy.cov = data.frame(pos = 1:4000, depth = lasy.depth[126:4125])

pdf("Fig4C.pdf", width = 28)
ggplot(Random.cov, aes(x=pos, y=depth))+
  geom_ribbon(data = Lasy.cov,aes(x=pos,ymin=0, ymax=depth), alpha = 0.5, fill = "#F8766D")+
  geom_ribbon(aes(ymin=0, ymax=depth), alpha = 0.5, fill = "#00BFC4")+
  scale_x_reverse()+
  xlab("Position from 3' end")+
  theme_bw()
dev.off()

#### Fig. 4C
cors.method = NULL
for(i in 0:5){
  cors.method = c(cors.method, cor(log2rpm.expressed[,49+i], log2rpm.expressed[,55+i]))
}

mean(cors.method)
sd(cors.method)

mean(cor.mat.SR[25:30,]$Peason.correlation)
sd(cor.mat.SR[25:30,]$Peason.correlation)

cor(log2rpm.expressed[,50], log2rpm.expressed[,49])
cor(log2rpm.expressed[,56], log2rpm.expressed[,55])
mean.sd.lasy1 = data.frame(method = "Lasy-Seq", mean = apply(log2rpm.expressed[,49:51],1,mean),sd = apply(log2rpm.expressed[,49:51],1,sd))
mean.sd.lasy2 = data.frame(method = "Lasy-Seq", mean = apply(log2rpm.expressed[,52:54],1,mean),sd = apply(log2rpm.expressed[,52:54],1,sd))
mean.sd.random1 = data.frame(method = "Conventional method", mean = apply(log2rpm.expressed[,55:57],1,mean),sd = apply(log2rpm.expressed[,55:57],1,sd))
mean.sd.random2 = data.frame(method = "Conventional method", mean = apply(log2rpm.expressed[,58:60],1,mean),sd = apply(log2rpm.expressed[,58:60],1,sd))
mean.sd.lasy = rbind(mean.sd.lasy1, mean.sd.lasy2)
mean.sd.random = rbind(mean.sd.random1, mean.sd.random2)
mean.sd = rbind(mean.sd.lasy, mean.sd.random)
library(gridExtra)

pdf("Fig4DE.pdf", width = 14)
g1 = ggplot(mean.sd.lasy, aes(x=mean, y = sd))+
  geom_point(alpha=0.05,color = "#F8766D")+
  theme_bw()+
  ylim(c(0,4.5))+
  ggtitle("Lasy-Seq")
g2 = ggplot(mean.sd.random, aes(x=mean, y = sd))+
  geom_point(alpha=0.05,color = "#00BFC4")+
  theme_bw()+
  ylim(c(0,4.5))+
  ggtitle("Conventional method")
grid.arrange(g1,g2,ncol = 2)
dev.off()

#### Fig. 4F 
at2.df = data.frame("method"= rep(c(rep("Lasy-Seq",6),rep("Conventional method",6)),5),"total.read"=at2[,2], "Num.of.detected.genes" =colSums(rawcnt.expressed>0))
pdf("Fig4F.pdf")
ggplot(at2.df, aes(x=total.read,y = Num.of.detected.genes, group= interaction(method,total.read)))+
  geom_boxplot(aes(fill= method))+
  theme_bw()+
  ylim(c(0,26000))
dev.off()

#### Fig. 4G
library(TCC)
FDR=0.05
num.DEG = NULL
DEG.list = list()
pdf("MAplot.pdf")
for(n in size){
  for(lib in libs){
    tmp = rawcnt.expressed[,(at2[,2]==n)&(is.element(at2[,1],at$V2[at$V3==lib]))]
    group = c(1,1,1,2,2,2)
    tcc <- new("TCC", tmp, group)       #TCCクラスオブジェクトtccを作成
    tcc <- calcNormFactors(tcc, norm.method="tmm", test.method="edger", iteration=3, FDR=0.1, floorPDEG=0.05)
    tcc <- estimateDE(tcc, test.method="edger", FDR=FDR)
    result <- getResult(tcc, sort=FALSE)
    plot(tcc)
    tmp = list(tcc$gene_id[tcc$estimatedDEG==1])
    names(tmp) = sprintf("%s-%s",lib,n)
    DEG.list = c(DEG.list, tmp)
    num.DEG = rbind(num.DEG,data.frame("lib size"=n, "method"=lib, "Num DEGs" = sum(tcc$stat$q.value < FDR)))
  }
}
dev.off()

library(ggplot2)
pdf("Fig4G.pdf")
ggplot(num.DEG, aes(x=lib.size,y=Num.DEGs,group=method))+
  geom_line(aes(color=method))+
  theme_bw()
dev.off()


#### Caluculation of Pearson's correlation between each library
cor.mat = NULL
for(lib in libs){
  for (n in size){
    cors = NULL
    tmp = log2rpm.expressed[,(at2[,2]==n)&(is.element(at2[,1],at$V2[at$V3==lib]))]
    for(a in 1:3){
      for (b in 1:3) {
        if(a<b){
          cors = cor(tmp[,a],tmp[,b])
          cor.mat = rbind(cor.mat,data.frame("method" = lib, "total-read" = as.integer(n), "Peason.correlation"= cors, "Index"=sprintf("%s_%s",a,b)))
        }
      }
    }
    for(a in 4:6){
      for (b in 4:6) {
        if(a<b){
          cors = cor(tmp[,a],tmp[,b])
          cor.mat = rbind(cor.mat,data.frame("method" = lib, "total-read" = as.integer(n), "Peason.correlation"= cors, "Index"=sprintf("%s_%s",a,b)))
        }
      }
    }
  }
}
